### PR TITLE
Implementation of the TypeScript into the project

### PR DESCRIPTION
**Pull request by lialina**
Implementing type checking in components and Redux files of the project using TypeScript and changing the files extension from .js, .jsx to .ts, .tsx.

### CODE REVIEW

File: src/components/ContactForm/ContactForm.tsx [hunk 1]
```diff
@@ -2,8 +2,9 @@ import React from 'react';
 import { useState } from 'react';
 import styles from './ContactForm.module.css';
 import { createShortId } from '../../serviсes/shortId';
+import ContactFormProps from '../../interfaces/ContactForm.interface';
```
Comment (kerrimov) [inline src/components/ContactForm/ContactForm.tsx]: move `ContactFormProps ` into this file

File: src/components/ContactForm/ContactForm.tsx [hunk 2]
```diff
@@ -12,17 +13,15 @@ export default function ContactForm({ onSubmit }) {
...
   const numberInputId = createShortId();
 
-  const handleSubmit = event => {
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
     event.preventDefault();
 
-    onSubmit({
-      contact,
-    });
+    onSubmit(contact,);
```
Comment (kerrimov) [inline src/components/ContactForm/ContactForm.tsx]: remove `,`

File: src/components/ContactList/ContactList.tsx [hunk 3]
```diff
@@ -1,16 +1,17 @@
 import React from 'react';
-import PropTypes from 'prop-types';
 import styles from './ContactList.module.css';
-import { useDispatch, useSelector } from 'react-redux';
+import { useSelector } from 'react-redux';
 import { deleteContact, getContacts } from '../../redux/actions';
 import { deleteContactService } from '../../serviсes/deleteContactService';
 import * as phonebookSelectors from '../../redux/selectors';
+import { useAppDispatch } from '../../redux/store'
+import ContactListProps from '../../interfaces/ContactList.interface';
```
Comment (kerrimov) [inline src/components/ContactList/ContactList.tsx]: move `ContactListProps ` into this file

Comment (lialina) [reply src/components/ContactList/ContactList.tsx]: All interfaces including this one have been moved to their own files.

File: src/components/ContactForm/ContactForm.tsx (hunk 2 above)
Comment (lialina) [reply src/components/ContactForm/ContactForm.tsx]: Removed.

File: src/components/ContactForm/ContactForm.tsx (hunk 1 above)
Comment (lialina) [reply src/components/ContactForm/ContactForm.tsx]: All interfaces including this one have been moved to their own files.

File: src/components/ContactForm/ContactForm.tsx [hunk 4]
```diff
@@ -3,7 +3,11 @@ import { useState } from 'react';
 import styles from './ContactForm.module.css';
 import { createShortId } from '../../serviсes/shortId';
 
-export default function ContactForm({ onSubmit }) {
+interface ContactFormProps {
+  onSubmit: Function
+}
+
+export default function ContactForm({ onSubmit }: ContactFormProps) {
   const [contact, setContact] = useState({
```
Comment (kerrimov) [inline src/components/ContactForm/ContactForm.tsx]: please provide type for this `useState`

File: src/react-app-env.d.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+/// <reference types="react-scripts" />
```
Comment (kerrimov) [inline src/react-app-env.d.ts]: remove this file from PR please

File: src/redux/store.ts [hunk 6]
```diff
@@ -8,3 +9,7 @@ export const store = configureStore({
   reducer: rootReducer,
   devTools: process.env.NODE_ENV === "development",
 })
+
+export type RootState = ReturnType<typeof store.getState>
```
Comment (kerrimov) [inline src/redux/store.ts]: Where `RootState` type is used?

File: src/redux/store.ts [hunk 7]
```diff
@@ -8,3 +9,7 @@ export const store = configureStore({
...
+
+export type AppDispatch = typeof store.dispatch
```
Comment (kerrimov) [inline src/redux/store.ts]: create types file inside redux folder and move this type here

File: src/components/ContactForm/ContactForm.tsx (hunk 4 above)
Comment (lialina) [reply src/components/ContactForm/ContactForm.tsx]: Type for `useState` has been provided as interface `ContactState`. I tried to use the `ContactData` type, but the editor shows an error that the `const contact` in the useState doesn't have an `id`.

File: src/react-app-env.d.ts (hunk 5 above)
Comment (lialina) [reply src/react-app-env.d.ts]: Removed.

File: src/redux/store.ts (hunk 6 above)
Comment (lialina) [reply src/redux/store.ts]: While not used anywhere, removed and will be added as needed.

File: src/redux/store.ts (hunk 7 above)
Comment (lialina) [reply src/redux/store.ts]: `types.ts` file has been created in `redux` folder and `AppDispatch` type was moved into it.

File: src/components/ContactForm/ContactForm.tsx (hunk 4 above)
Comment (kerrimov) [reply src/components/ContactForm/ContactForm.tsx]: remove `ContactState` and use `ContactData`. Use optional field in this case

File: src/App.tsx [hunk 8]
```diff
@@ -7,10 +7,12 @@ import { useDispatch, useSelector } from 'react-redux';
...
+import type { AppDispatch } from './redux/types'
 
 export default function App() {
-  const dispatch = useDispatch();
-  const [filter, setFilter] = useState('');
+  const dispatch = useDispatch<AppDispatch>();
+  const [filter, setFilter] = useState<string>('');
 
   const items = useSelector(phonebookSelectors.items);
   const loader = useSelector(phonebookSelectors.loader);
```
Comment (kerrimov) [inline src/App.tsx]: 1. Why types for `useSelector` have not been provided?
2. State as parameter should be provided

File: src/redux/store.ts (hunk 6 above)
Comment (kerrimov) [reply src/redux/store.ts]: Should not be removed, return and use in useSelector

File: src/components/ContactList/ContactList.tsx [hunk 9]
```diff
@@ -1,16 +1,21 @@
...
 
-export default function ContactList({ contactsData }) {
-  const dispatch = useDispatch();
+interface ContactListProps {
+  contactsData: ContactData[]
+}
+
+export default function ContactList({ contactsData }: ContactListProps) {
+  const dispatch = useDispatch<AppDispatch>();
   const error = useSelector(phonebookSelectors.error);
```
Comment (kerrimov) [inline src/components/ContactList/ContactList.tsx]: 1. Why types for useSelector have not been provided?
2. State as parameter should be provided

File: src/App.tsx [hunk 10]
```diff
@@ -33,14 +35,14 @@ export default function App() {
...
-  const changeFilter = event => {
+  const changeFilter = (event: React.ChangeEvent<HTMLInputElement>) => {
     setFilter(event.currentTarget.value);
   };
 
   const getVisibleContacts = useMemo(() => {
     const normalizedFilter = filter.toLowerCase();
 
-    let visibleContacts = items.filter(contact =>
+    let visibleContacts = items.filter((contact: ContactData) =>
```
Comment (kerrimov) [inline src/App.tsx]: Why `let`?

File: src/components/ContactList/ContactList.tsx (hunk 9 above)
Comment (lialina) [reply src/components/ContactList/ContactList.tsx]: Type for useSelector in ContactList has been provided.

**Review by kerrimov — APPROVED** [review]

**Pull request merged**